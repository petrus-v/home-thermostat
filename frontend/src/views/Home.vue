<template>
  <div data-se="home-page">
    <section>
      <div class="field">
          <b-switch 
              :disabled="isDesiredBurnerLoading"
              v-on:input="onChangeDesiredBurnerState"
              v-model="burnerDesiredState"
              type='is-warning'
              size="is-large"
              passive-type='is-dark'>
              Etat désiré du bruleur
              <b-loading :is-full-page="false" v-model="isDesiredBurnerLoading" :can-cancel="false">
              </b-loading>
          </b-switch>
          <b-switch 
              :disabled="true"
              v-model="burnerState"
              type='is-warning'
              size="is-large"
              passive-type='is-dark'>
              Etat du bruleur
              <b-loading :is-full-page="false" v-model="isBurnerLoading" :can-cancel="false">
              </b-loading>
          </b-switch>
      </div>
      <div class="field">
          <b-switch 
              :disabled="isDesiredEngineLoading"
              v-on:input="onChangeDesiredEngineState"
              v-model="engineDesiredState"
              type='is-success'
              size="is-large"
              passive-type='is-dark'>
              Etat désiré du circulateur
              <b-loading :is-full-page="false" v-model="isDesiredEngineLoading" :can-cancel="false">
              </b-loading>
          </b-switch>
          <b-switch 
              :disabled="true"
              v-model="engineState"
              type='is-warning'
              size="is-large"
              passive-type='is-dark'>
              Etat du circulateur
              <b-loading :is-full-page="false" v-model="isEngineLoading" :can-cancel="false">
              </b-loading>
          </b-switch>
      </div>
      <div class="buttons">
          <b-button
            @click="getStates"
            icon-left="refresh">
              Rafraîchir
          </b-button>
      </div>
    </section>
  </div>
</template>

<script>
export default {
  name: "Home",
  data() {
    return {
      isBurnerLoading: true,
      isDesiredBurnerLoading: true,
      burnerState: null,
      burnerDesiredState: null,
      isEngineLoading: true,
      isDesiredEngineLoading: true,
      engineState: null,
      engineDesiredState: null,
    }
  },
  methods: {
    getStates() {
      this.getBurnerState();
      this.getBurnerDesiredState();
      this.getEngineState();
      this.getEngineDesiredState();
    },
    getBurnerState() {
      this.getOrSetState(
        "GET",
        "BURNER",
        "isBurnerLoading",
        "burnerState",
        null
      );
    },
    onChangeDesiredBurnerState(new_value) {
      this.getOrSetState(
        "POST",
        "BURNER/desired",
        "isDesiredBurnerLoading",
        "burnerDesiredState",
        JSON.stringify({ is_open: !new_value })
      );
      // TODO: something weird happens in backend
      // if(new_value &&  !this.engineDesiredState){
      //   this.onChangeDesiredEngineState(new_value);
      // }
    },
    getBurnerDesiredState() {
      this.getOrSetState(
        "GET",
        "BURNER/desired",
        "isDesiredBurnerLoading",
        "burnerDesiredState",
        null
      );
    },
    getEngineState() {
      this.getOrSetState(
        "GET",
        "ENGINE",
        "isEngineLoading",
        "engineState",
        null
      );
    },
    onChangeDesiredEngineState(new_value) {
      this.getOrSetState(
        "POST",
        "ENGINE/desired",
        "isDesiredEngineLoading",
        "engineDesiredState",
        JSON.stringify({ is_open: !new_value })
      );
    },
    getEngineDesiredState() {
      this.getOrSetState(
        "GET",
        "ENGINE/desired",
        "isDesiredEngineLoading",
        "engineDesiredState",
        null
      );
    },
    getOrSetState(method, device, loaderName, stateName, payload) {
      this[loaderName] = true;
      let self = this;
      fetch(`/api/device/${device}/state`, {
        method: method,
        body: payload,
        headers: { 
          "Content-Type": "application/json"
        }
      })
        .then(function(response) {
          if (!response.ok) {
            self[loaderName] = true;
            self[stateName] = null;
            throw new Error(`HTTP error! status: ${response.status}`);
          }
          if (response.ok) return response.json();
        })
        .then(function(state) {
          self[stateName] = !state.is_open;
          self[loaderName] = false;
        })
        .catch(err => {
          console.error("Error: ", err);
          self[loaderName] = true;
          self[stateName] = null;
        });
    }
  },
  mounted() {
    this.getStates();
  },
};
</script>
