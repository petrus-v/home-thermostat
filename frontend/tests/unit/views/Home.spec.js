import { mount } from "@vue/test-utils";
import Home from "@/views/Home";
import flushPromises from "flush-promises";
const localVue = global.localVue;
const store = global.store;

const API_RESULTS = {
  "/api/device/ENGINE/desired/state": Promise.resolve({ is_open: true }),
  "/api/device/BURNER/desired/state": Promise.resolve({ is_open: true }),
  "/api/device/ENGINE/state": Promise.resolve({ is_open: true }),
  "/api/device/BURNER/state": Promise.resolve({ is_open: false }),
};

global.fetch = jest.fn((api, content) => {
  return Promise.resolve({
    ok: true,
    json: () => API_RESULTS[api],
  });
});

describe("Home", () => {
  beforeEach(() => {
    fetch.mockClear();
  });

  it("mount Home view", async () => {
    const wrapper = mount(Home, {
      store,
      localVue,
    });
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot();
  });
});
