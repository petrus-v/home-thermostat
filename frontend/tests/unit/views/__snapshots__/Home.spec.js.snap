// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home mount Home view 1`] = `
<div
  data-se="home-page"
>
  <section>
    <div
      class="field"
    >
      <label
        class="switch is-large is-rounded"
      >
        <input
          true-value="true"
          type="checkbox"
          value="false"
        />
        <span
          class="check is-dark-passive is-warning"
        />
        <span
          class="control-label"
        >
          
        Etat désiré du bruleur
        
          <transition-stub
            name="fade"
          >
            <!---->
          </transition-stub>
        </span>
      </label>
       
      <label
        class="switch is-large is-disabled is-rounded"
        disabled="disabled"
      >
        <input
          disabled="disabled"
          true-value="true"
          type="checkbox"
          value="false"
        />
        <span
          class="check is-dark-passive is-warning"
        />
        <span
          class="control-label"
        >
          
        Etat du bruleur
        
          <transition-stub
            name="fade"
          >
            <!---->
          </transition-stub>
        </span>
      </label>
    </div>
     
    <div
      class="field"
    >
      <label
        class="switch is-large is-rounded"
      >
        <input
          true-value="true"
          type="checkbox"
          value="false"
        />
        <span
          class="check is-dark-passive is-success"
        />
        <span
          class="control-label"
        >
          
        Etat désiré du circulateur
        
          <transition-stub
            name="fade"
          >
            <!---->
          </transition-stub>
        </span>
      </label>
       
      <label
        class="switch is-large is-disabled is-rounded"
        disabled="disabled"
      >
        <input
          disabled="disabled"
          true-value="true"
          type="checkbox"
          value="false"
        />
        <span
          class="check is-dark-passive is-warning"
        />
        <span
          class="control-label"
        >
          
        Etat du circulateur
        
          <transition-stub
            name="fade"
          >
            <!---->
          </transition-stub>
        </span>
      </label>
    </div>
     
    <div
      class="buttons"
    >
      <button
        class="button"
        type="button"
      >
        <span
          class="icon is-small"
        >
          <i
            class="mdi mdi-refresh"
          />
        </span>
        <span>
          Rafraîchir
        </span>
        <!---->
      </button>
    </div>
  </section>
</div>
`;
